cub/test/catch2_test_device_select_if.cu

```diff
@@ -0,0 +1,300 @@
+
+  out.resize(num_selected_out[0]);
+  reference.resize(num_selected_out[0]);
+  REQUIRE(reference == out);
+}
+
+CUB_TEST("DeviceSelect::If works with iterators", "[device][select_if]", all_types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  // just pick one of the input elements as boundary
+  less_than_t<type> le{in[num_items / 2]};
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_first_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  select_if(in.begin(),
+            out.begin(),
+            d_first_num_selected_out,
+            num_items,
+            le);
+
+  const auto boundary = out.begin() + num_selected_out[0];
+  REQUIRE(thrust::all_of(out.begin(), boundary, le));
+  REQUIRE(thrust::all_of(boundary, out.end(), equal_to_default_t{}));
+}
+
+CUB_TEST("DeviceSelect::If works with pointers", "[device][select_if]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  // just pick one of the input elements as boundary
+  less_than_t<type> le{in[num_items / 2]};
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_first_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  select_if(thrust::raw_pointer_cast(in.data()),
+            thrust::raw_pointer_cast(out.data()),
+            d_first_num_selected_out,
+            num_items,
+            le);
+
+  const auto boundary = out.begin() + num_selected_out[0];
+  REQUIRE(thrust::all_of(out.begin(), boundary, le));
+  REQUIRE(thrust::all_of(boundary, out.end(), equal_to_default_t{}));
+}
+
+CUB_TEST("DeviceSelect::If works in place", "[device][select_if]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  // just pick one of the input elements as boundary
+  less_than_t<type> le{in[num_items / 2]};
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_first_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  // Ensure that we create the same output as std
+  thrust::host_vector<type> reference = in;
+  std::stable_partition(reference.begin(), reference.end(), le);
+
+  select_if(in.begin(),
+            d_first_num_selected_out,
+            num_items,
+            le);
+
+  in.resize(num_selected_out[0]);
+  reference.resize(num_selected_out[0]);
+  REQUIRE(reference == in);
+}
```

Comment:
Is there anything that corresponds to old `TestMixedOp` in this test?

Reply:
We discussed this offline and my opinion is that those tests only covered other thrust functionality. The only thing that related to the actual input type was that the struct was convertible to bool, which we have a special test for.

Reply:
> The only thing that related to the actual input type was that the struct was convertible to bool
I believe this is about normalization test. The `TestMixedOp` tests that input and output iterator can have different value types.

Reply:
You are right, I added a specific test for that